### PR TITLE
[VPlan] Handle exit phis with multiple operands in addUsersInExitBlocks.

### PR DESCRIPTION
**Pull request by fhahn**
Currently the addUsersInExitBlocks incorrectly assumes exit phis only have a single operand, which may not be the case for loops with early exits when they share a common exit block.

Also further relax the assertion in fixupIVUsers to allow exit values if they come from theloop latch/middle.block.

### CODE REVIEW

**Comment by david-arm** [comment]
Hi @fhahn, is this patch just fixing an existing bug? As part of #88385 I have added support for dealing with live-outs from early exit loops and I have patches downstream that I plan to put upstream as soon as #117008 lands.

**Comment by fhahn** [comment]
yes this fixes an existing bug, i.e. cases where we incorrectly would try to vectorize an early exit loop with non-live-in exit values. The particular case we missed is a single exit phi with a live-in incoming value from the loop latch and a non-live-in from the early exit;

Previously we only checked the first operand of the VPIRInstruction wrapping the exit phi and would only check the live-in.

Now as a side effect this particular case we can now handle the inverse case (live-in from the early exit and IV from the latch) unless I am missing anything, as we should be able to compute the exit value from the latch as previously and pass through the live-in.

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp
Comment (david-arm) [inline llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:9052]: Should this be the index corresponding to the predecessor instead of 0? It doesn't matter for this patch I think, but I suppose it just looks a little odd that's all!

**Review by david-arm — APPROVED** [review]
LGTM! I left one comment about possibly using an index instead of 0, but wouldn't hold up the patch for it.

File: llvm/lib/Transforms/Vectorize/LoopVectorize.cpp
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/LoopVectorize.cpp]: Yep updated to `enumerate`, thanks

**Review by david-arm — APPROVED** [review]
LGTM!

**Pull request merged**
